zerver/webhooks/newrelic/view.py

```diff
@@ -1,169 +1,264 @@
+                        ]
+                    )
+                )
+            )
+            if isinstance(details, list):
+                custom_field_detail_str = ", ".join([f"**`{detail}`**" for detail in details])
+            else:
+                custom_field_detail_str = f"**{details}**"
+
+            custom_field_message = f"|{custom_field_name_str}: {custom_field_detail_str}|\n"
+            body_custom_field_detail += custom_field_message
+        except ValidationError:
+            invalid_field_message = f"|*The **{custom_field_name_str}** contains [unsupported field](/integrations/doc/newrelic#displaying-custom-fields) format.* |\n"
+            body_custom_field_detail += invalid_field_message
+    return body_custom_field_detail
+
+
+@webhook_view("NewRelic", all_event_types=INCIDENT_STATES)
 @typed_endpoint
 def api_newrelic_webhook(
     request: HttpRequest,
     user_profile: UserProfile,
     *,
     payload: JsonBodyPayload[WildValue],
 ) -> HttpResponse:
-    # Handle old format
-    # Once old is EOLed, delete if block and keep else block
-    if not payload.get("id").tame(check_none_or(check_string_fixed_length(36))):
-        info = {
-            "condition_name": payload.get("condition_name", "Unknown condition").tame(check_string),
-            "details": payload.get("details", "No details.").tame(check_string),
-            "incident_url": payload.get("incident_url", "https://alerts.newrelic.com").tame(
-                check_string
-            ),
-            "incident_acknowledge_url": payload.get(
-                "incident_acknowledge_url", "https://alerts.newrelic.com"
-            ).tame(check_string),
-            "status": payload.get("current_state", "None").tame(check_string),
-            "iso_timestamp": "",
-            "owner": payload.get("owner", "").tame(check_string),
-        }
-
-        unix_time = payload.get("timestamp").tame(
-            check_none_or(check_union([check_string, check_int]))
-        )
-        if unix_time is None:
-            raise JsonableError(_("The newrelic webhook requires timestamp in milliseconds"))
-
-        info["iso_timestamp"] = str(unix_milliseconds_to_timestamp(unix_time, "newrelic"))
-
-        # Add formatting to the owner field if owner is present
-        if info["owner"] != "":
-            info["owner"] = "by **{}** ".format(info["owner"])
-
-        # These are the three promised current_state values
-        if info["status"].lower() == "open":
-            content = OPEN_TEMPLATE.format(**info)
-        elif info["status"].lower() == "acknowledged":
-            content = DEFAULT_TEMPLATE.format(**info)
-        elif info["status"].lower() == "closed":
-            content = DEFAULT_TEMPLATE.format(**info)
-        else:
-            raise JsonableError(
-                _("The newrelic webhook requires current_state be in [open|acknowledged|closed]")
-            )
+    is_complete = check_essential_fields(payload)
+    if is_complete:
```

Comment:
Can't this condition just be the function? And we could name it better for this use. Something like "payload_has_expected_fields" ...

```python
    if payload_has_expected_fields(payload):
        message_context = parse_payload(payload)
```